File: two-sum-linearithmic-time/TwoSum.java
// https://leetcode.com/problems/two-sum
// LC1: Two Sum
// Time/space: Linearithmic time, linear extra-space.
// Concepts: Binary Search.
class TwoSum {
    int find(Item[] items, int start, int target) {
        int lo = start;
        int hi = items.length-1;
        while(lo < hi) {
            int mid = (lo+hi)/2;
            if (target <= items[mid].num) {
                hi = mid;
            } else {
                lo = mid+1;
            }
        }
        return items[lo].num == target? lo: -1;
    }
    class Item implements Comparable<Item> {
        int id;        
        int num;
        Item(int id, int num) {
            this.id = id;
            this.num = num;
        }
        @Override
        public int compareTo(Item other) {
            return Integer.compare(this.num, other.num);
        }
    }
    public int[] twoSum(int[] nums, int target) {
        int n = nums.length;
        Item[] items = new Item[n];
        for (int i=0; i<n; i++) {
            items[i] = new Item(i, nums[i]);
        }
        Arrays.sort(items);
        for (int i=0; i<n-1; i++) {
            int complement = target-items[i].num;
            int j = find(items, i+1, complement); 
            if (j != -1) {
                return new int[] {items[i].id, items[j].id};
            }
        }
        return new int[] {};
    }
}

File: two-sum/two-sum-linearithmic-time-v1/TwoSum.cpp
// https://leetcode.com/problems/two-sum
// Linearithmic time, linear extra-space.
// Concept: Sorting, Binary search, Symmetry, Data-index pairing.
class TwoSum {
private:
    struct item {
        int num;
        int idx;
    };
    int search(vector<item>& items, int lo, int hi, int target) {
        while (lo < hi) {
            int mid = (lo+hi)/2;
            int num = items[mid].num;
            if (target <= num) {
                hi = mid;
            } else {
                lo = mid+1;
            }
        }
        int num = items[lo].num;
        return (num == target)? lo: -1;
    }
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int n = nums.size();
        vector<item> items(n);
        for (int i=0; i<n; i++) { 
            items[i] = {nums[i], i};
        }
        sort(items.begin(), items.end(), [](item lhs, item rhs) {
            return lhs.num < rhs.num;
        });
        for (int i=0; i<n-1; i++) {
            int num = items[i].num;
            int j = search(items, i+1, n-1, target-num);
            if (j != -1) {
                return {items[i].idx, items[j].idx};
            }
        }
        return {};
    }
};
